runtime/commonMain/src/kotlinx/serialization/internal/CollectionSerializers.kt

```diff
@@ -140,6 +140,45 @@ sealed class MapLikeSerializer<TKey, TVal, TCollection, TBuilder: MutableMap<TKe
     }
 }
 
+abstract class PrimitiveArraySerializer<TElement, TCollection, TBuilder : PrimitiveArraySerializer<TElement, TCollection, TBuilder>.Builder>
+internal constructor(
+    primitiveSerializer: KSerializer<TElement>,
+    primitiveDescriptor: PrimitiveDescriptor
+) : ListLikeSerializer<TElement, TCollection, TBuilder>(primitiveSerializer) {
+    final override val descriptor = PrimitiveArrayDescriptor(primitiveDescriptor)
+
+    abstract inner class Builder internal constructor() {
+        abstract val pos: Int
+        abstract fun ensureCapacity(requiredCapacity: Int = pos + 1)
```

Comment:
All array parsed can be accidentally quadratic. Please use `position * 2` explicitly. I've seen it is used in each particular implementation, this is too error-prone. Additionally, such behaviour (`coerceAtLeast(size * 2)`) prevents precise sizing when size is known in advance (e.g. for "fixed size array" type)

Reply:
It would not be quadratic due to the `coerceAtLeast` part, would it? If size is known in advance and was set during `ensureCapacity(readSize())`, then `if (buf.size < requiredCapacity)` would never be true and resizing would not be called

Reply:
Well, it would be true in one case: when INITIAL_SIZE = 10 and `readSize()` returned something from 10 to 20 (say, 15), we'll still allocate 20 slots. I'm not sure that this is a big concern though.

Reply:
> Please use position * 2 explicitly.

That will cause resize on each append, which you correctly noticed later in comments